In [25]:
# import all the libraries
import pandas as pd
import random
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from sklearn.metrics import classification_report, confusion_matrix, multilabel_confusion_matrix
from sklearn.metrics import mean_squared_error, accuracy_score, precision_score, recall_score
from sklearn.model_selection import cross_validate


In [6]:
# import updated dataset
updatedBC = pd.read_csv('updatedDataset.csv')

# taking X and y in the dataset
X_updatedBC = updatedBC.drop('diagnosis', axis = 1)
y_updatedBC = updatedBC['diagnosis']

# print the dataset
print(X_updatedBC)
print(updatedBC['diagnosis'].unique())
print("Dimensions of the dataset :", updatedBC.shape)

     Unnamed: 0  radius_mean  texture_mean  perimeter_mean  area_mean  \
0             0        17.99         10.38          122.80     1001.0   
1             1        20.57         17.77          132.90     1326.0   
2             2        19.69         21.25          130.00     1203.0   
3             3        11.42         20.38           77.58      386.1   
4             4        20.29         14.34          135.10     1297.0   
..          ...          ...           ...             ...        ...   
551         563        20.92         25.09          143.00     1347.0   
552         564        21.56         22.39          142.00     1479.0   
553         565        20.13         28.25          131.20     1261.0   
554         566        16.60         28.08          108.30      858.1   
555         567        20.60         29.33          140.10     1265.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760      

In [7]:
# data normalization X
scaler = MinMaxScaler(feature_range=(0, 1))
X_rescaled = scaler.fit_transform(X_updatedBC)
X_normalized = pd.DataFrame(data = X_rescaled, columns = X_updatedBC.columns)

# encoding y
set_of_classes = y_updatedBC.value_counts().index.tolist()
set_of_classes= pd.DataFrame({'diagnosis': set_of_classes})
y_encoded = pd.get_dummies(y_updatedBC)

print(X_normalized)
print(y_encoded)

     Unnamed: 0  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      0.000000     0.504383      0.022658        0.531250   0.356389   
1      0.001764     0.630736      0.272574        0.603311   0.495838   
2      0.003527     0.587639      0.390260        0.582620   0.443062   
3      0.005291     0.182624      0.360839        0.208619   0.092551   
4      0.007055     0.617023      0.156578        0.619007   0.483395   
..          ...          ...           ...             ...        ...   
551    0.992945     0.647877      0.520122        0.675371   0.504849   
552    0.994709     0.679220      0.428813        0.668236   0.561486   
553    0.996473     0.609188      0.626987        0.591182   0.467948   
554    0.998236     0.436309      0.621238        0.427797   0.295074   
555    1.000000     0.632205      0.663510        0.654680   0.469664   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0           0.553970          0.792037      

In [8]:
# set up trying parameters for hyperparameter tuning
max_iterations = [500, 600, 700, 800, 900, 1000]
hidden_layer_siz = []
for i in range(1, 31):
    hidden_layer_siz.append((i, 30))
for i in range(1, 31):
    hidden_layer_siz.append((30, i))
batch_size = 25 * np.arange(1, 10)
learning_rates = 0.05 * np.arange(1, 10)
random_num = random.randint(0, 100000)


param_grid = dict(learning_rate_init = learning_rates, hidden_layer_sizes = hidden_layer_siz, max_iter = max_iterations, batch_size = batch_size)

# model
mlp = MLPClassifier(solver = 'sgd', random_state = random_num, activation = 'logistic', learning_rate_init = 0.15, batch_size = 30, hidden_layer_sizes = (12, 3), max_iter = 500)

# grid search
grid = GridSearchCV(estimator = mlp, param_grid = param_grid)

# random search
random = RandomizedSearchCV(estimator = mlp, param_distributions = param_grid, n_iter = 10)

In [9]:
grid.fit(X_normalized, y_encoded)

/Users/amon/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/amon/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/amon/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/amon/anaconda3/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/amon/anaconda3/lib/python3.11/sit

GridSearchCV(estimator=MLPClassifier(activation='logistic', batch_size=30,
                                     hidden_layer_sizes=(12, 3),
                                     learning_rate_init=0.15, max_iter=500,
                                     random_state=98262, solver='sgd'),
             param_grid={'batch_size': array([ 25,  50,  75, 100, 125, 150, 175, 200, 225]),
                         'hidden_layer_sizes': [(3, 6), (4, 7), (5, 8), (6, 9),
                                                (7, 10), (8, 11), (9, 12),
                                                (10, 13), (11, 14), (12, 15),
                                                (13, 16), (14, 17), (15, 18),
                                                (16, 19), (17, 20), (18, 21),
                                                (19, 22), (20, 23), (21, 24),
                                                (22, 25), (23, 26), (24, 27),
                                                (25, 28), (26, 29), (27, 30),
                                                (28, 31), (29, 32), (30, 33),
                                                (31, 34), (32, 3), ...],
                         'learning_rate_init': array([0.05, 0.1 , 0.15, 0.2 , 0.25, 0.3 , 0.35, 0.4 , 0.45]),
                         'max_iter': [500, 550]})

In [10]:
print("random: ", random_num)
print("Optimal Hyper-parameters : ", grid.best_params_)
print("Optimal Accuracy : ", grid.best_score_)

random:  98262
Optimal Hyper-parameters :  {'batch_size': 225, 'hidden_layer_sizes': (25, 28), 'learning_rate_init': 0.45, 'max_iter': 500}
Optimal Accuracy :  0.976657014157014


|seed| batch size | parameters | accuracy|
|--| -- | -- | --|
|114514| 30 | {'hidden_layer_sizes': (16, 5), 'learning_rate_init': 0.09, 'max_iter': 500} | 0.976657014157014 |
|1919810| 30 |   {'hidden_layer_sizes': (6, 9), 'learning_rate_init': 0.09, 'max_iter': 500} |  0.9766409266409267 |
||30|{'hidden_layer_sizes': (3, 6), 'learning_rate_init': 0.18, 'max_iter': 500}|0.9766409266409267|
|29898|30|{'hidden_layer_sizes': (7, 10), 'learning_rate_init': 0.3, 'max_iter': 500}|0.9766409266409267|
|65046|30|{'hidden_layer_sizes': (3, 6), 'learning_rate_init': 0.39, 'max_iter': 500}|0.9766248391248391|
|76959|30|{'hidden_layer_sizes': (16, 6), 'learning_rate_init': 0.44999999999999996, 'max_iter': 500}|0.9784427284427284|
|56392|30|{'hidden_layer_sizes': (16, 4), 'learning_rate_init': 0.06, 'max_iter': 500}|0.978426640926641|
|22952||{'hidden_layer_sizes': (16, 8), 'learning_rate_init': 0.42, 'max_iter': 500}|0.9766409266409266|
|61563||{'hidden_layer_sizes': (16, 5), 'learning_rate_init': 0.36, 'max_iter': 500}|0.9766409266409267|
|28120||{'hidden_layer_sizes': (4, 7), 'learning_rate_init': 0.42, 'max_iter': 500}|0.9766409266409267|

(16,6)

In [35]:
chosen_mlp =  MLPClassifier(solver = 'sgd', random_state = random.randint(0, 100000)
, activation = 'logistic', learning_rate_init = 0.45, batch_size = 225, hidden_layer_sizes = (25, 28), max_iter = 500)

In [40]:
CV = cross_validate(chosen_mlp, X_normalized, y_encoded, cv=10, scoring=['accuracy', 'neg_mean_squared_error'])

print('Accuracy')
print(CV['test_accuracy'])
print('MSE')
print(-1*CV['test_neg_mean_squared_error'])

print('Average Accuracy = ', sum(CV['test_accuracy']) / len(CV['test_accuracy']))
print('Average MSE = ', sum(-1 * CV['test_neg_mean_squared_error']) / len(CV['test_neg_mean_squared_error']))

Accuracy
[0.96428571 0.94642857 0.92857143 0.96428571 0.94642857 0.96428571
 0.98181818 0.98181818 1.         0.96363636]
MSE
[0.03571429 0.05357143 0.07142857 0.03571429 0.05357143 0.03571429
 0.01818182 0.01818182 0.         0.03636364]
Average Accuracy =  0.964155844155844
Average MSE =  0.035844155844155845


In [41]:
X_train, X_test, y_train, y_test = train_test_split(X_normalized, y_encoded, test_size=0.2)

chosen_mlp.fit(X_train, y_train)
pred = chosen_mlp.predict(X_test)
print(pred)

[[0 1]
 [0 1]
 [0 1]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [0 1]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [0 1]
 [0 1]
 [1 0]
 [1 0]
 [0 1]
 [0 1]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [0 1]
 [1 0]
 [0 1]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [0 1]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [0 1]
 [1 0]
 [0 1]
 [0 1]
 [1 0]
 [0 1]
 [1 0]
 [1 0]
 [1 0]
 [0 1]
 [0 1]
 [1 0]
 [0 1]
 [0 1]
 [1 0]
 [1 0]
 [1 0]]


In [42]:
print("Accuracy : ", accuracy_score(y_test, pred))
print("Mean Square Error : ", mean_squared_error(y_test, pred))

print(pred[:5])
print("Confusion Matrix for each label : ")
print(multilabel_confusion_matrix(y_test, pred))

print("Classification Report : ")
print(classification_report(y_test, pred))

Accuracy :  0.9821428571428571
Mean Square Error :  0.017857142857142856
[[0 1]
 [0 1]
 [0 1]
 [1 0]
 [1 0]]
Confusion Matrix for each label : 
[[[37  0]
  [ 2 73]]

 [[73  2]
  [ 0 37]]]
Classification Report : 
              precision    recall  f1-score   support

           0       1.00      0.97      0.99        75
           1       0.95      1.00      0.97        37

   micro avg       0.98      0.98      0.98       112
   macro avg       0.97      0.99      0.98       112
weighted avg       0.98      0.98      0.98       112
 samples avg       0.98      0.98      0.98       112

